frontend/annotation_resolver.py

```diff
@@ -60,82 +35,138 @@ def resolve(self, annotation, solver):
-            solver.add(Or([result_type == arg for arg in args_types]),
-                       fail_message="Union in type annotation")
-
+        if isinstance(annotation, ast.Name):
+            if annotation.id in self.primitives:
+                return self.primitives[annotation.id]
+            if annotation.id in self.z3_types.all_types:
+                return getattr(self.z3_types.type_sort, "class_{}".format(annotation.id))
+            
+            # Check if it's a generic type var
+            if generics_map is None:
+                raise ValueError("Invalid type annotation {} in line {}".format(annotation.id, annotation.lineno))
+            if annotation.id in generics_map:
+                return generics_map[annotation.id]
+            if annotation.id not in self.type_vars:
+                raise ValueError("Invalid type annotation {} in line {}".format(annotation.id, annotation.lineno))
+
+            result_type = solver.new_z3_const("generic")
+            generics_map[annotation.id] = result_type
+
+            supers_types = [self.resolve(x, solver, generics_map) for x in self.type_vars[annotation.id]]
+            if supers_types:
+                solver.add(Or([solver.z3_types.subtype(result_type, x) for x in supers_types]),
+                           fail_message="Generic type in line {}".format(annotation.lineno))
             return result_type
 
-        raise ValueError("Invalid type annotation: {}.".format(annotation))
+        if isinstance(annotation, ast.Subscript):
+            assert isinstance(annotation.value, ast.Name)
+            assert isinstance(annotation.slice, ast.Index)
+            annotation_val = annotation.value.id
+            if annotation_val == "List":
+                # Parse List type
+                return self.z3_types.list(self.resolve(annotation.slice.value, solver, generics_map))
+            
+            if annotation_val == "Dict":
+                # Parse Dict type
+                assert isinstance(annotation.slice.value, ast.Tuple)
+                assert len(annotation.slice.value.elts) == 2
+
+                # Get the types of the dict args
+                keys_type = self.resolve(annotation.slice.value.elts[0], solver, generics_map)
+                vals_type = self.resolve(annotation.slice.value.elts[1], solver, generics_map)
+                return self.z3_types.dict(keys_type, vals_type)
+            
+            if annotation_val == "Set":
+                # Parse Set type
+                return self.z3_types.set(self.resolve(annotation.slice.value, solver, generics_map))
+            
+            if annotation_val == "Type":
+                # Parse Type type
+                return self.z3_types.type(self.resolve(annotation.slice.value, solver, generics_map))
+            
+            if annotation_val == "Tuple":
+                # Parse Tuple type
+                assert isinstance(annotation.slice.value, (ast.Name, ast.Tuple))
```

Comment:
Won't fail. A tuple with a single value will have an instance of `Name` as its `annotation.slice.value`, which passes the assertion.

To have an empty tuple, you can write `Tuple[()]`. I looked for the way PEP 484 does it but I didn't find any useful info.